go/ql/src/experimental/CWE-525/WebCacheDeception.ql

```diff
@@ -0,0 +1,29 @@
+/*
+ * @name Web Cache Deception
+ * @description A caching system has been detected on the application and is vulnerable to web cache deception. By manipulating the URL it is possible to force the application to cache pages that are only accessible by an authenticated user. Once cached, these pages can be accessed by an unauthenticated user.
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 9
+ * @precision high
+ * @id go/web-cache-deception
+ * @tags security
+ *       external/cwe/cwe-525
+ */
+
+import go
+
+from
+  DataFlow::CallNode httpHandleFuncCall, DataFlow::CallNode call, DataFlow::Node predecessor,
+  Method get
+where
+  httpHandleFuncCall.getTarget().hasQualifiedName("net/http", "HandleFunc") and
+  httpHandleFuncCall.getNumArgument() > 1 and
+  httpHandleFuncCall.getArgument(0).getType().toString() = "string" and
+  httpHandleFuncCall.getArgument(0).toString().matches("%/\"") and
+  // Trace the second argument's data flow to its predecessor
+  predecessor = httpHandleFuncCall.getArgument(1).getAPredecessor() and
+  // Find the corresponding expression for the predecessor
+  get.hasQualifiedName("net/http", "Header", "Set") and
+  call = get.getACall() and
+  call.getArgument(0).toString().matches("\"Cache-Control\"")
+select httpHandleFuncCall.getArgument(0), call.getArgument(0)
```

Comment:
This is not the right format for a path problem. I will try to find some documentation to show you what it should be.

Reply:
https://codeql.github.com/docs/writing-codeql-queries/creating-path-queries/#constructing-a-path-query
But actually, this isn't a path problem, as it's currently written.

Reply:
Currently this query checks that (1) `httpHandleFuncCall` is a call to `HandleFunc` and its first argument ends in "/", and (2) `call` is a call to `Header.Set` and its first argument is "Cache-Control". What it doesn't do it relate those two things in any way. These two calls could be in totally different files, and not relate to the same objects. I'm not totally sure how you want to relate them. Perhaps the next paragraph will help.

I looked for where `Header.Set` is already used in our code and found `private class HeaderWriteCall` in `go/ql/lib/semmle/go/frameworks/stdlib/NetHttp.qll`. It's private so you can't use it directly. But it extends `Http::HeaderWrite::Range`, so you can use `Http::HeaderWrite` and get it (and many more classes which do similar things in other libraries). It is defined [here](https://github.com/github/codeql/blob/803ed209629ac74bddef82be36b19f708e97cbfd/go/ql/lib/semmle/go/concepts/HTTP.qll#L99). If you aren't familiar with "the range pattern" this might be bit confusing, but basically try using `HeaderWrite` and it should just work. Do you think you could `HeaderWrite` instead?

Reply:
To be clear, I'm expecting something like "there exists a HeaderWrite with getHeaderName() = A and getHeaderValue = B" plus something linking the headerwrite to httpHandleFuncCall.

Reply:
I use HeaderWrite::Range in following syntax.

`hw.getHeaderName() = "cache-control"`

Reply:
Yes. In my case I am checking is Cache Control used in the following function which used in the endpoint but I don't think that this HeaderWrite::Range check the following function. I think first method decrease the False Positive percentage.